### DISCUSSION TITLE
Basic use of lifecycle functions

### DISCUSSION THREAD

**Discussion by joejackson3** [arkivanov/Essenty#152] [Q&A]
This is a very promising library. So far, I haven't been able to figure out how to use it for my Kotlin Multiplatform project. 
My only real requirement is to run a OnResume function to update my main activity, when it is brought back into focus from background.

I am not clear where to instantiate the functions:
```
val lifecycleRegistry = LifecycleRegistry()
val someLogic = SomeLogic(lifecycleRegistry)

```
Do I need to call `lifecycleRegistry.subscribe()` from the Composable of the activity in question?

What I'm currently doing in my main module is define the SomeLogic class: 
```
class SomeLogic(lifecycle: Lifecycle,t:MutableState<String>) {
    init {
        lifecycle.subscribe(
            onResume = {t.value="Res1"}
        )
        lifecycle.doOnResume {
             t.value = "Res";
        }
    }
}
```
then instantiate the members in the Composable of the main activity:
The OnResume function is however never called.
```
val lifecycleRegistry = LifecycleRegistry()
val someLogic = SomeLogic(lifecycleRegistry,t3)
lifecycleRegistry.subscribe()
lifecycleRegistry.doOnCreate { t3.value="yo" }
```
The OnResume function is however never called.
Any advice or reference to sample projects would be highly appreciated.

**Comment by arkivanov** [comment]
For Android, you can use one of the Android extensions provided by Essenty (as [mentioned](https://github.com/arkivanov/Essenty?tab=readme-ov-file#android-extensions) in the readme). For example, you can do the following in Activity.

```kotlin
class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val lifecycle = essentyLifecycle()
        val someLogic = SomeLogic(lifecycle = lifecycle)
    }
}
```

**Reply by joejackson3** [reply]
Thanks. I got this to work in Android, but how I can use it in Multiplatform, i.e. commonMain.kotlin module of my KMP project?

**Reply by arkivanov** [reply]
So `Lifecycle` interface is available in `commonMain`, so you can use that in KMP. Then you should bind the `Lifecycle` on platform separately, because each platform manages the lifecycle differently. The general approach is to create `LifecycleRegistry` on the platform side, subscribe to lifecycle events provided by the platform, and drive `LifecycleRegistry` accordingly.

On iOS, this can be done in two ways:

- Create `LifecycleRegistry` on Swift side and subscribe to lifecycle events depending on your use case. It can be events from `UIViewController` or [application lifecycle events](https://developer.apple.com/documentation/uikit/uiapplicationdelegate#1656902).
- The as above but in Kotlin, see [ApplicationLifecycle](https://github.com/arkivanov/Decompose/blob/8d56415e219ccd8d8dbf68d5fe58c7695daa6add/decompose/src/itvosMain/kotlin/com/arkivanov/decompose/lifecycle/ApplicationLifecycle.kt#L30) as example. I will move `ApplicationLifecycle` to Essenty so that it could be reused.

With Compose for Desktop (JVM) - we can use events from `WindowState`, see [example](https://github.com/arkivanov/Decompose/blob/master/extensions-compose-jetbrains/src/jvmMain/kotlin/com/arkivanov/decompose/extensions/compose/jetbrains/lifecycle/LifecycleController.kt).

On Web (Kotlin/JS) - use `document` events, see [example](https://github.com/arkivanov/Decompose/blob/8d56415e219ccd8d8dbf68d5fe58c7695daa6add/sample/app-js/src/main/kotlin/com/arkivanov/sample/app/Main.kt#L40).

**Reply by joejackson3** [reply]
Thank you very much. Will have a look into that.

**Reply by joejackson3** [reply]
I will look forward to seeing ApplicationLifecycle being integrated into Essenty, then I'll try again. Thanks

**Reply by arkivanov** [reply]
Will be available in the next alpha release.

**Reply by arkivanov** [reply]
Essenty 2.0.0-alpha07 is released! See https://github.com/arkivanov/Essenty/releases/tag/2.0.0-alpha07

**Reply by arkivanov** [reply]
Also see the update readme: https://github.com/arkivanov/Essenty?tab=readme-ov-file#ios-and-tvos-extensions-since-v200-alpha07
